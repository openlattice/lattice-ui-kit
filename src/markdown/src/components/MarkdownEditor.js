// @flow
import React, { useEffect, useRef, useState } from 'react';

import styled from 'styled-components';
import { Tab, Tabs } from '@material-ui/core';
import type { TextFieldProps } from '@material-ui/core';

import MarkdownPreview from './MarkdownPreview';

import TextArea from '../../../text/src/components/TextArea';

const StyledTabPanel = styled.div`
  display: ${(props) => (props.show ? 'block' : 'none')};
  padding: 8px 0 0;
`;

const StyledTextArea = styled(TextArea)`
  textarea {
    min-height: 84px;
  }
`;

const EDIT = 'edit';
const PREVIEW = 'preview';

type ViewType = 'edit' | 'preview';

type Props = {
  defaultValue :?string;
  value :string;
  view :ViewType;
  ...TextFieldProps;
};

const MarkdownEditor = ({
  defaultValue,
  onChange = () => {},
  value,
  view = EDIT,
  ...rest
} :Props) => {
  const [content, setContent] = useState(value || defaultValue);
  const [tab, setTab] = useState(view);
  const didMount = useRef(false);
  const inputRef = useRef();

  useEffect(() => {
    if (didMount.current && tab === EDIT && inputRef.current) {
      inputRef.current.focus();
    }
    else {
      didMount.current = true;
    }
  }, [inputRef, tab]);

  const handleChangeTab = (event, newValue) => {
    if (newValue === EDIT && inputRef.current) {
      inputRef.current.focus();
    }
    setTab(newValue);
  };

  const handleChangeContent = (e) => {
    onChange(e);
    setContent(e.target.value);
  };

  return (
    <div>
      <Tabs
          aria-label="markdown editor tabs"
          indicatorColor="primary"
          onChange={handleChangeTab}
          textColor="primary"
          value={tab}>
        <Tab label="Edit" value={EDIT} />
        <Tab label="Preview" value={PREVIEW} />
      </Tabs>
      <StyledTabPanel show={tab === EDIT}>
        <StyledTextArea
            inputRef={inputRef}
            onChange={handleChangeContent}
            rows={4}
            value={content}
            /* eslint-disable-next-line react/jsx-props-no-spreading */
            {...rest} />
      </StyledTabPanel>
      <StyledTabPanel show={tab === PREVIEW}>
        <MarkdownPreview>{content}</MarkdownPreview>
      </StyledTabPanel>
    </div>
  );
};

export default MarkdownEditor;
export {
  StyledTabPanel
};
