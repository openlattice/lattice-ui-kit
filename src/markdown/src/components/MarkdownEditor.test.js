import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import MarkdownEditor from './MarkdownEditor';

describe('MarkdownEditor', () => {

  test('render matches snapshot', () => {
    const wrapper = shallow(<MarkdownEditor />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
