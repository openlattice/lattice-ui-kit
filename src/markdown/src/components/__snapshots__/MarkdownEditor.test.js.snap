// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownEditor render matches snapshot 1`] = `
<MarkdownEditor
  forwardedRef={null}
/>
`;
